### PR TITLE
GTEST: Added truncation check for contig, generic and iov datatype tests

### PR DESCRIPTION
**Pull request by alinask**

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2982/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/499/ for details.

File: test/gtest/ucp/test_ucp_tag_xfer.cc [hunk 1]
```diff
@@ -244,19 +249,26 @@ void test_ucp_tag_xfer::test_xfer_generic(size_t size, bool expected, bool sync)
...
+        truncated = false;
+    }
+
     status = ucp_dt_create_generic(&test_dt_uint32_ops, this, &dt);
     ASSERT_UCS_OK(status);
 
-    recvd = do_xfer(NULL, NULL, count, dt, dt, expected, sync, false);
-    EXPECT_EQ(count * sizeof(uint32_t), recvd);
+    recvd = do_xfer(NULL, NULL, count, dt, dt, expected, sync, truncated);
+    if (!truncated) {
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_tag_xfer.cc]: we would want to check that in case of truncated, recd<=(count*sizeof(uint32_t)

File: test/gtest/ucp/test_ucp_tag_xfer.cc [hunk 2]
```diff
@@ -244,19 +249,26 @@ void test_ucp_tag_xfer::test_xfer_generic(size_t size, bool expected, bool sync)
     dt_gen_start_count  = 0;
     dt_gen_finish_count = 0;
 
+    if ((truncated) && (!count)) {
+        truncated = false;
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_tag_xfer.cc]: why is that needed?

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2446/ for details (Mellanox internal link).

File: test/gtest/ucp/test_ucp_tag_xfer.cc (hunk 2 above)
Comment (alinask) [inline test/gtest/ucp/test_ucp_tag_xfer.cc]: because if count is calculated to zero, the truncation has no effect.

Comment (yosefe) [inline test/gtest/ucp/test_ucp_tag_xfer.cc]: I see. So in general, it's a bad practice to change function arguments, since it may lead to all sorts of confusion. It would be better to change the 'if' below to:
`if (!truncated || (count==0)) {`

**Comment by Di0gen** [comment]
mlx:bot:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2989/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/518/ for details.

File: test/gtest/ucp/test_ucp_tag_xfer.cc (hunk 2 above)
Comment (alinask) [inline test/gtest/ucp/test_ucp_tag_xfer.cc]: but this would set 'truncated' to false when it's already false

Comment (yosefe) [inline test/gtest/ucp/test_ucp_tag_xfer.cc]: i meant - not assign anything to 'truncated' at all. just change the 'if' in line 260 to check the value of 'count' in addition to 'truncated'.

Comment (alinask) [inline test/gtest/ucp/test_ucp_tag_xfer.cc]: the status of the rreq is checked inside do_xfer, so changing the 'if' in line 260 would still result in the test's failure.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/520/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2454/ for details (Mellanox internal link).

File: test/gtest/ucp/test_ucp_tag_xfer.cc (hunk 2 above)
Comment (yosefe) [inline test/gtest/ucp/test_ucp_tag_xfer.cc]: ic, ok then

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2992/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/524/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/526/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2458/ for details (Mellanox internal link).
